src/ekam/rules/intercept.c

```diff
@@ -285,6 +286,103 @@ static void canonicalizePath(char* path) {
   }
 }
 
+static bool path_has_prefix(const char* pathname, const char* prefix, size_t prefix_length) {
+  // Returns true if pathname is a child within prefix or if it matches prefix directly. Prefix must
+  // always end with a trailing '/'.
+  assert(prefix_length > 1);
+  assert(prefix_length <= PATH_MAX);
+  assert(prefix[prefix_length - 1] == '/');
+  if (strncmp(pathname, prefix, prefix_length - 1) == 0) {
+    return pathname[prefix_length - 1] == '\0' || pathname[prefix_length - 1] == '/';
+  }
+
+  return false;
+}
+
+static bool bypass_remap(const char *pathname) {
+  int debug = EKAM_DEBUG;
+
+  if (pathname[0] != '/') {
+    // Only absolute paths are allowed to bypass the remap.
+    return false;
+  }
+
+  char* bypass_dirs = getenv("EKAM_REMAP_BYPASS_DIRS");
+  if (bypass_dirs == NULL) {
+    return false;
+  }
+
+  bool found = false;
+
+  bypass_dirs = strdup(bypass_dirs);
+  // Duplicated here because I want to avoid any concurrency risk with multiple threads modifying
```

Comment:
Why would another thread be modifying the environment variable?

If another thread _is_ modifying it, what stops it from doing the modification between when this thread runs `getenv()` and when it runs `strdup`, thus still causing a problem?

Reply:
If I modify the result of `getenv` in-situ, then I have multiple threads modifying it potentially. If I duplicate then I'm always processing a local copy & so modifications of the string is harmless.

Reply:
Oh I see, you are modifying the string content below... I didn't quite catch that.

But why are you modifying the content? `path_has_prefix` takes a length and doesn't appear to care if the string is NUL-terminated, so it seems like you don't actually need to replace the separators with NULs?

Reply:
It's perhaps subtle but it does require the string to be NUL-terminated.

```
return pathname[prefix_length - 1] == '\0' || pathname[prefix_length - 1] == '/';
```

The first part of that conditional would be false. I guess I could check for `:` there as well on the assumption that no file we'll encounter in practice has `:` in the name. That would obviate the `strdup` need.

Reply:
AFAICT, the line you quoted is checking `pathname`, whereas `prefix` is the string that derives from `bypass_dirs`. So the line you quoted doesn't appear to be affected by writing NULs into `bypass_dirs`?